services/src/pro/datasets/postgres.rs

```diff
@@ -861,11 +912,631 @@ impl From<FileUpload> for crate::datasets::upload::FileUpload {
+                    display_name,
+                    description,
+                    source_operator,
+                    result_descriptor,
+                    meta_data,
+                    symbology,
+                    provenance,
+                    tags
+                )
+                VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10::text[])",
+            &[
+                &id,
+                &name,
+                &dataset.display_name,
+                &dataset.description,
+                &dataset.source_operator,
+                &typed_meta_data.result_descriptor,
+                typed_meta_data.meta_data,
+                &dataset.symbology,
+                &dataset.provenance,
+                &dataset.tags,
+            ],
+        )
+        .await
+        .map_unique_violation("datasets", "name", || error::Error::InvalidDatasetName)?;
+
+        let stmt = tx
+            .prepare(
+                "
+            INSERT INTO permissions (
+                role_id,
+                dataset_id,
+                permission
+            )
+            VALUES ($1, $2, $3)",
+            )
+            .await?;
+
+        tx.execute(
+            &stmt,
+            &[&RoleId::from(self.session.user.id), &id, &Permission::Owner],
+        )
+        .await?;
+
+        let stmt = tx
+            .prepare(
+                "
+            INSERT INTO uploaded_user_datasets (
+                user_id,
+                upload_id,
+                dataset_id,
+                status,
+                created,
+                expiration,
+                deleted,
+                delete_data,
+                delete_record
+            )
+            VALUES ($1, $2, $3, 'Available', CURRENT_TIMESTAMP, NULL, NULL, false, false)",
+            )
+            .await?;
+
+        tx.execute(
+            &stmt,
+            &[&RoleId::from(self.session.user.id), &upload_id, &id],
+        )
+        .await?;
+
+        tx.commit().await?;
+
+        Ok(DatasetIdAndName { id, name })
+    }
+
+    async fn expire_uploaded_dataset(&self, expire_dataset: ChangeDatasetExpiration) -> Result<()> {
+        let mut conn = self.conn_pool.get().await?;
+        let tx = conn.build_transaction().start().await?;
+
+        self.ensure_permission_in_tx(expire_dataset.dataset_id.into(), Permission::Owner, &tx)
+            .await
+            .boxed_context(crate::error::PermissionDb)?;
+
+        self.update_dataset_status_in_tx(&tx, &expire_dataset.dataset_id)
+            .await?;
+
+        match expire_dataset.expiration_change {
+            ExpirationChange::SetExpire(expiration) => {
+                let num_changes = if let Some(delete_timestamp) = expiration.deletion_timestamp {
+                    let stmt = tx
+                        .prepare("
+                        UPDATE uploaded_user_datasets
+                        SET status = 'Expires', expiration = $2, delete_data = $3, delete_record = $4
```

Comment:
maybe set the "Expires" as a parameter with a Rust enum variant instead of a string? Or is it fine like this?

Reply:
Not sure if this is actually more readable? Or did you have a format! with an inplace Enum in mind (if that works)?  7b4fa9c2b0233db64d3505387f34555d02c2b069